### PR TITLE
Copter terrain following (3rd attempt)

### PR DESCRIPTION
**Pull request by rmackay9**
Third attempt at adding terrain following and absolute altitude handling to Copter.

The way it works (not changed from previous PR):
- set TERRAIN_USE parameter to "1" to enable the use of terrain data for LAND and RTL modes
- AUTO and GUIDED modes always accept terrain altitudes (regardless of the TERRAIN_USE parameter's value) if the command's MAV_FRAME is set to MAV_FRAME_GLOBAL_TERRAIN_ALT (see http://mavlink.org/messages/common)

This version resolves these issues found in the previous PR:
1. removed complicated extrapolation code in favour of a terrain failsafe which triggers if terrain data is not available
2. reduces use of Location_Class's get_alt() to access terrain data for an arbitrary location and instead directly calls the terrain library's height_above_terrain function.
3. renamed TERRAIN_USE to TERRAIN_FOLLOW to be consistent with Plane

We have the following issues remaining (to be added in a follow-up PR):
c) do-set-roi does not yet accept terrain altitudes

### CODE REVIEW

File: libraries/AP_Common/Location.cpp [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        Location ekf_origin;
+        if (_ahrs->get_origin(ekf_origin)) {
+            lat = ekf_origin.lat;
+            lng = ekf_origin.lng;
+            offset(ekf_offset_neu.x / 100.0f, ekf_offset_neu.y / 100.0f);
+        }
+    }
+}
+
+Location_Class& Location_Class::operator=(const struct Location loc)
```
Comment (tridge) [inline libraries/AP_Common/Location.cpp]: this should take a const reference to prevent copying data too much

File: ArduCopter/Parameters.cpp [hunk 2]
```diff
@@ -960,6 +960,14 @@ const AP_Param::Info Copter::var_info[] = {
     // @User: Standard
     GSCALAR(throw_motor_start, "THROW_MOT_START", 0),
 
+    // @Param: TERRAIN_FOLLOW
+    // @DisplayName: Terrain Following use control
+    // @Description: Control if terrain data is used in RTL and LAND
+    // @Description: This enables terrain following for RTL and LAND flight modes. To use this option TERRAIN_ENABLE must be 1 and the GCS must  support sending terrain data to the aircraft.  In RTL the RTL_ALT will be considered a height above the terrain.  In LAND mode the vehicle will slow to LAND_SPEED 10m above terrain (instead of 10m above home).  This parameter does not affect AUTO and Guided which use a per-command flag to determine if the height is above-home, absolute or above-terrain.
```
Comment (OXINARF) [inline ArduCopter/Parameters.cpp]: Is the double _Description_ intended?

File: ArduCopter/control_auto.cpp [hunk 3]
```diff
@@ -400,37 +449,64 @@ void Copter::auto_rtl_run()
 // auto_circle_movetoedge_start - initialise waypoint controller to move to edge of a circle with it's center at the specified location
 //  we assume the caller has set the circle's circle with circle_nav.set_center()
```
Comment (OXINARF) [inline ArduCopter/control_auto.cpp]: This comment is no longer true right?

File: ArduCopter/control_auto.cpp [hunk 4]
```diff
@@ -400,37 +449,64 @@ void Copter::auto_rtl_run()
...
 //  we assume the caller has performed all required GPS_ok checks
-void Copter::auto_circle_movetoedge_start()
+void Copter::auto_circle_movetoedge_start(const Location_Class &circle_center, float radius_m)
 {
+    // convert location to vector from ekf origin
+    Vector3f circle_center_neu;
+    if (!circle_center.get_vector_from_origin_NEU(circle_center_neu)) {
+        // default to current position and log error
+        circle_center_neu = inertial_nav.get_position();
```
Comment (OXINARF) [inline ArduCopter/control_auto.cpp]: Comment says log error but there's no error logging

File: libraries/AP_Terrain/AP_Terrain.cpp [hunk 5]
```diff
@@ -143,22 +147,30 @@ bool AP_Terrain::height_amsl(const Location &loc, float &height)
         home_loc = loc;
     }
 
+    // apply correction which assumes home altitude is at terrain altitude
+    if (corrected) {
+        height += (ahrs.get_home().alt * 0.01f) - home_height;
+    }
```
Comment (OXINARF) [inline libraries/AP_Terrain/AP_Terrain.cpp]: Two points about this:

- the _corrected_ argument is only used in one place so I'm not sure if it is valuable to have it in the function (the caller could do it)
- I'm not understanding what this does. If home altitude is at terrain altitude then _home_height_ wouldn't be the same as _ahrs.get_home().alt_?

Comment (rmackay9) [inline libraries/AP_Terrain/AP_Terrain.cpp]: Thanks OXNIARF, I think you've found a bug.  I think height_terrain_difference_home should also be setting that to true.  What's going on here is we are assuming that the terrain altitude is zero at home (i.e. that the vehicle is taking off from the terrain).  The way we do that, for better or worse, is we essentially adjust the terrain up and down.

**Comment by rmackay9** [comment]
Ok, I've fixed up nearly all the issues found by Tridge and @OXINARF.  I didn't find a bug in the "correct" parameter of the height_amsl() method.  I haven't moved the altitude correction out of height_amsl (as pointed out by @OXINARF's) but I plan to in a future commit.

**Pull request merged**